File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/map/submodelelement/ReferenceElement.java
package org.eclipse.basyx.submodel.metamodel.map.submodelelement;

import java.util.Map;

import org.eclipse.basyx.submodel.metamodel.api.reference.IReference;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.IReferenceElement;
import org.eclipse.basyx.submodel.metamodel.map.modeltype.ModelType;
import org.eclipse.basyx.submodel.metamodel.map.reference.Reference;
import org.eclipse.basyx.submodel.metamodel.map.submodelelement.property.Property;

/**
 * A ReferenceElement as defined in DAAS document <br/>
 * A reference element is a data element that defines a reference to another
 * element within the same or another AAS or a reference to an external object
 * or entity.
 * 
 * @author schnicke, pschorn
 *
 */
public class ReferenceElement extends DataElement implements IReferenceElement {
	public static final String MODELTYPE = "ReferenceElement";

	/**
	 * Constructor
	 */
	public ReferenceElement() {
		// Add model type
		putAll(new ModelType(MODELTYPE));

		put(Property.VALUE, null);
	}

	/**
	 * @param ref
	 *            Reference to any other referable element of the same or any other
	 *            AAS or a reference to an external object or entity
	 */
	public ReferenceElement(Reference ref) {
		// Add model type
		putAll(new ModelType(MODELTYPE));
		put(Property.VALUE, ref);
	}
	
	/**
	 * Creates a ReferenceElement object from a map
	 * 
	 * @param obj a ReferenceElement object as raw map
	 * @return a ReferenceElement object, that behaves like a facade for the given map
	 */
	public static ReferenceElement createAsFacade(Map<String, Object> obj) {
		ReferenceElement ret = new ReferenceElement();
		ret.setMap(obj);
		return ret;
	}

	@Override
	public void setValue(IReference ref) {
		put(Property.VALUE, ref);
		
	}

	@Override
	@SuppressWarnings("unchecked")
	public IReference getValue() {
		return Reference.createAsFacade((Map<String, Object>) get(Property.VALUE));
	}

}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/map/submodelelement/ReferenceElement.cpp
/*
 * ReferenceElement.cpp
 *
 *      Author: wendel
 */

#include <BaSyx/submodel/map/submodelelement/ReferenceElement.h>
#include <BaSyx/submodel/map/reference/Reference.h>

namespace basyx {
namespace submodel {

ReferenceElement::ReferenceElement() :
  vab::ElementMap{},
  ModelType{IReferenceElement::Path::Modeltype}
{}

ReferenceElement::ReferenceElement(const IReference & reference) :
  vab::ElementMap{},
  ModelType{IReferenceElement::Path::Modeltype}
{
  this->setValue(reference);
}

ReferenceElement::ReferenceElement(const basyx::object & map) :
  vab::ElementMap{map}
{}

void ReferenceElement::setValue(const IReference & ref)
{
  Reference reference {ref};
  this->insertMapElement(IProperty::Path::Value, reference);
}

std::shared_ptr<IReference> ReferenceElement::getValue() const
{
  return std::make_shared<Reference>(this->map.getProperty(IProperty::Path::Value));
}

}
}
